portal/index.php

```diff
@@ -27,15 +27,58 @@
 
 //includes
 require_once('../interface/globals.php');
+require_once(dirname(__FILE__) . "/lib/appsql.class.php");
+$logit = new ApplicationTable();
 
+use OpenEMR\Common\Crypto\CryptoGen;
 use OpenEMR\Core\Header;
 
 //exit if portal is turned off
 if (!(isset($GLOBALS['portal_onsite_two_enable'])) || !($GLOBALS['portal_onsite_two_enable'])) {
     echo xlt('Patient Portal is turned off');
     exit;
 }
-
+$auth['portal_pwd'] = '';
+if (isset($_GET['woops'])) {
+    unset($_GET['woops']);
+    unset($_SESSION['password_update']);
+}
+if (isset($_GET['forward'])) {
+    $auth = false;
+    if (strlen($_GET['forward']) >= 64) {
+        $crypto = new CryptoGen();
+        if ($crypto->cryptCheckStandard($_GET['forward'])) {
+            $one_time = $crypto->decryptStandard($_GET['forward'], null, 'drive', 6);
+            if (!empty($one_time)) {
+                $auth = sqlQueryNoLog("Select * From patient_access_onsite Where portal_onetime Like BINARY ?",
+                    array(($one_time . '%')));
+            }
+        }
+    }
+    if ($auth === false) {
+        error_log("PORTAL ERROR: " . errorLogEscape('One time reset:' . $_GET['forward']), 0);
+        $logit->portalLog('login attempt', '', ($_GET['forward'] . ':invalid one time'), '', '0');
+        OpenEMR\Common\Session\SessionUtil::portalSessionCookieDestroy();
+        header('Location: ' . $landingpage . '&w&u');
+        exit();
+    }
+    $parse = str_replace($one_time, '', $auth['portal_onetime']);
+    $validate = hex2bin(substr($parse, 6));
+    if ($validate <= time()) {
+        error_log("PORTAL ERROR: " . errorLogEscape('One time reset link expired. Dying.'), 0);
+        $logit->portalLog('password reset attempt', '', ($_POST['uname'] . ':link expired'), '', '0');
+        OpenEMR\Common\Session\SessionUtil::portalSessionCookieDestroy();
+        die(xlt("Your one time credential reset link has expired. Reset and try again.") . "time:$validate time:" . time());
+    }
+    $_SESSION['pin'] = substr($parse, 0, 6);
+    $_SESSION['forward'] = $auth['portal_onetime'];
+    $_SESSION['portal_username'] = $auth['portal_username'];
+    $_SESSION['portal_login_username'] = '';
+    $_SESSION['password_update'] = 2;
+    $_SESSION['onetime'] = $auth['portal_pwd'];
+    sqlStatementNoLog("UPDATE `patient_access_onsite` SET `portal_onetime`=NULL WHERE BINARY `portal_onetime` = ?", [$auth['portal_onetime']]);
```

Comment:
yep, the above line was a bad idea by me :)
(I basically broke the token since the sql entry is needed...)

Reply:
maybe just clear on new credential save in get_patient_info